fuzz/writeout.c

```diff
@@ -0,0 +1,216 @@
+        default:
+            return "InvalidUnknownValue";
+    }
+}
+
+static bool time_to_str(long time, char** buf, size_t* buf_size)
+{
+    time /= 10; // ASS files can only have centi-second precision
+    int sign = time < 0 ? -1 : 1;
+    time = labs(time);
+
+    // Without chars for hours, except for sign indicator
+    size_t min_size = sign == -1 ? 14 : 10;
+    int cs, s, m;
+    #define DIVMOD(div,mod) mod = time % div; time /= div
+    DIVMOD(100, cs);
+    DIVMOD(60, s);
+    DIVMOD(60, m);
+    #undef DIVMOD
+
+    min_size += time >= 10 ? ceil(log10(time + 1)) : 1;
+    if (*buf_size < min_size) {
+        // +14 are arbitrary padding to avoid too many reallocs
+        char* nbuf = realloc(*buf, min_size + 14);
+        if (!nbuf) {
+            return false;
+        }
+        *buf = nbuf;
+        *buf_size = min_size + 14;
+    }
+
+    int ret = snprintf(*buf, *buf_size, "%ld:%02d:%02d.%02d",
+                       time * sign, m * sign, s * sign, cs * sign);
+    if (ret < 0 || ret >= *buf_size)
+        return false;
+    return true;
+}
+
+static int ssa2ass_align(int ssa_align)
+{
+    return ((ssa_align & 0xC) >> 2) * 3 + (ssa_align & 0x3);
+}
+
+static const char *tracktype_to_str(int track_type)
+{
+    return track_type == TRACK_TYPE_ASS ? "ASS"
+        : track_type == TRACK_TYPE_SSA ? "SSA"
+        : "other";
+}
+
+
+static void write_header(FILE *f, ASS_Track *track, const char* originalformat_name)
+{
+    // format version is normalised to ASS
+    fprintf(f, "[Script Info]\n");
+    fprintf(f, "; Original Format: %s\n", originalformat_name);
+    fprintf(f, "ScriptType: v4.00+\n");
+
+    #define HEADER_INT(name)  fprintf(f, "%s: %d\n",   #name, track->name)
+    #define HEADER_FLT(name)  fprintf(f, "%s: %.3f\n", #name, track->name)
+    #define HEADER_STR(name)  if (track->name) fprintf(f, "%s: %s\n",   #name, track->name)
+    #define HEADER_BOOL(name) fprintf(f, "%s: %s\n",   #name, track->name ? "yes" : "no")
+
+    HEADER_INT(PlayResX);
+    HEADER_INT(PlayResY);
+    HEADER_INT(LayoutResX);
+    HEADER_INT(LayoutResY);
+    HEADER_FLT(Timer);
+    HEADER_INT(WrapStyle);
+    HEADER_BOOL(ScaledBorderAndShadow);
+    HEADER_BOOL(Kerning);
+    HEADER_STR(Language);
+    if (track->YCbCrMatrix != YCBCR_DEFAULT) /* Or normalise this to TV.601? */
+        fprintf(f, "YCbCrMatrix: %s\n", ycbcr_to_str(track->YCbCrMatrix));
+
+    #undef HEADER_INT
+    #undef HEADER_FLT
+    #undef HEADER_STR
+    #undef HEADER_BOOL
+}
+
+static void write_styles(FILE *f, ASS_Track *track)
+{
+    fprintf(f, "\n[V4+ Styles]\n");
+    fprintf(f, "Format: "
+        "Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, "
+        "OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, "
+        "ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, "
+        "Alignment, MarginL, MarginR, MarginV, Encoding\n");
+
+    // don't print out builtin fallback
+    for (int i = 1; i < track->n_styles; ++i) {
```

Comment:
I think we decided to use `i++` in this context.

Reply:
fixed here and at another place